# Intro

This will be an example of data analysis on data of data analyst jobs from the website glassdoor.

I'm aware I might not be using the most efficient ways of doing things (I specialise more in the deep data science side of things: deep learning, etc), to some advanced programmers this might make their eyes bleed, however I'm using this as an end to end project (without the scraping, I'm saving that for another project) to learn from, if you have any advice I would appreciate it if you shared it with me wherever that's via issues on the github page or a comments section. Thanks!

This will be seperated into chapters where various transformations and various other methods will be applied/used;

**[Insights](#Insights)** = Insights gathered after all the transformations in the chapters below. If you are curious how all these were generated then read the later chapters to see the **Python code** used.

**[Chapter 1](#Chapter-1)** = Simple transformations of string data into appropriate integer data. 

**[Chapter 2](#Chapter-2)** = Some form of **Regex (Regular Expressions)** done in my own way (not familar with the Regex module yet). 


-------------------------------------------------------------------------------------------------------------------------------

# Insights
<a id='Insights'></a>

-------------------------------------------------------------------------------------------------------------------------------

# Chapter 1
<a id='Chapter 1'></a>



Import necessary Packages

In [83]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re

Load original data from CSV and view it and start planning transoformations to get it in a useable format.

In [84]:
pd.read_csv('C:/Users/Asus/Desktop/python/Data/jobs.csv')

Unnamed: 0                                          Job Title  \
0              0  Data Analyst, Center on Immigration and Justic...   
1              1                               Quality Data Analyst   
2              2  Senior Data Analyst, Insights & Analytics Team...   
3              3                                       Data Analyst   
4              4                             Reporting Data Analyst   
...          ...                                                ...   
2248        2248  RQS - IHHA - 201900004460 -1q Data Security An...   
2249        2249              Senior Data Analyst (Corporate Audit)   
2250        2250  Technical Business Analyst (SQL, Data analytic...   
2251        2251                Data Analyst 3, Customer Experience   
2252        2252                        Senior Quality Data Analyst   

                  Salary Estimate  \
0      $37K-$66K (Glassdoor est.)   
1      $37K-$66K (Glassdoor est.)   
2      $37K-$66K (Glassdoor est.)   
3      $37K-$66K (Glassdoor est.)   
4      $37K-$66K (Glassdoor est.)   
...                           ...   
2248  $78K-$104K (Glassdoor est.)   
2249  $78K-$104K (Glassdoor est.)   
2250  $78K-$104K (Glassdoor est.)   
2251  $78K-$104K (Glassdoor est.)   
2252  $78K-$104K (Glassdoor est.)   

                                        Job Description  Rating  \
0     Are you eager to roll up your sleeves and harn...     3.2   
1     Overview\n\nProvides analytical and technical ...     3.8   
2     We’re looking for a Senior Data Analyst who ha...     3.4   
3     Requisition NumberRR-0001939\nRemote:Yes\nWe c...     4.1   
4     ABOUT FANDUEL GROUP\n\nFanDuel Group is a worl...     3.9   
...                                                 ...     ...   
2248  Maintains systems to protect data from unautho...     2.5   
2249  Position:\nSenior Data Analyst (Corporate Audi...     2.9   
2250  Title: Technical Business Analyst (SQL, Data a...    -1.0   
2251  Summary\n\nResponsible for working cross-funct...     3.1   
2252  You.\n\nYou bring your body, mind, heart and s...     3.4   

                                 Company Name        Location  \
0              Vera Institute of Justice\n3.2    New York, NY   
1     Visiting Nurse Service of New York\n3.8    New York, NY   
2                            Squarespace\n3.4    New York, NY   
3                               Celerity\n4.1    New York, NY   
4                                FanDuel\n3.9    New York, NY   
...                                       ...             ...   
2248                       Avacend, Inc.\n2.5      Denver, CO   
2249                   Arrow Electronics\n2.9  Centennial, CO   
2250                                 Spiceorb      Denver, CO   
2251         Contingent Network Services\n3.1  Centennial, CO   
2252                          SCL Health\n3.4  Broomfield, CO   

          Headquarters                    Size  Founded  \
0         New York, NY    201 to 500 employees     1961   
1         New York, NY        10000+ employees     1893   
2         New York, NY  1001 to 5000 employees     2003   
3           McLean, VA    201 to 500 employees     2002   
4         New York, NY   501 to 1000 employees     2009   
...                ...                     ...      ...   
2248    Alpharetta, GA     51 to 200 employees       -1   
2249    Centennial, CO        10000+ employees     1935   
2250                -1                      -1       -1   
2251  West Chester, OH    201 to 500 employees     1984   
2252    Broomfield, CO        10000+ employees     1864   

                   Type of ownership                                 Industry  \
0             Nonprofit Organization                        Social Assistance   
1             Nonprofit Organization         Health Care Services & Hospitals   
2                  Company - Private                                 Internet   
3     Subsidiary or Business Segment                              IT Services   
4       

Assign loaded dataframe.

In [85]:
df = pd.read_csv('C:/Users/Asus/Desktop/python/Data/jobs.csv')

Drop unnecessary columns with the keyword inplace so we dont have to assign the new dataframes created.

In [86]:
df.drop('Easy Apply', axis=1, inplace=True)
df.drop('Competitors', axis=1, inplace=True)
df.drop('Headquarters', axis=1, inplace=True)
df.drop('Unnamed: 0', axis=1, inplace=True)

Convert the 'k' in 'salary estimate' column to thousands and get rid of unncessary characters, then convert range to the median value (remember it is in USD), then convert USD to £'s (today's rate : $1 = £0.76), try and except to capture potential nuls or anything else that isnt what we expect and delete the offending row (Be sure to check the offending rows beforehand! In this case it was null values denoted with a '-1' as shown below in the output).

In [87]:
for j, i in enumerate(df['Salary Estimate']):
        i = i.replace('K', '000')
        i = i.replace('(Glassdoor est.)', '')
        i = i.replace('$', '')
        i = i.strip('')
        indexofdash = i.find('-')
        num1 = i[:indexofdash]
        num2 = i[indexofdash+1:]
        try:
            new = int(((int(num1) + int(num2))/2)*0.76)
            df['Salary Estimate'][j] = new
        except ValueError:
            print(f"row {j} cannot be parsed as an int, it's value is {i}; dropping row")
            df.drop(j)

C:\ProgramData\Anaconda\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()


row 2149 cannot be parsed as an int, it's value is -1; dropping row


Convert column to be of 'integer' type rather than 'object', then use keyword downcast to convert to lowest memory-form of the integer datatype to save memory (within reason for the values in the column, which pandas decides).

In [88]:
df['Salary Estimate'] = pd.to_numeric(df['Salary Estimate'], downcast='integer')

Make new column for age of company (better for machine learning than year founded if we are going that route), does this for all values that have 4 characters else it marks it with a '?' so I can do some further digging into what the values are apart from years

In [89]:
df['Company Age'] = [2020 - i if len(str(i)) == 4 else '?' for i in df['Founded']]

Check the values that were not 4 characters in df['Founded']

In [90]:
values = {}
for j, i in enumerate(df['Company Age']):
    if i == '?':
        if df['Founded'][j] not in values.keys():
            values[df['Founded'][j]] = 1
        else:
            values[df['Founded'][j]] += 1
print(values)

{-1: 660}


All the values that were not what we expected are -1's which denote nulls, since 660 rows is too much to drop we will set the nulls as the mean of the companies age's, after we are going to drop the 'Founded' column as we no longer need it 

In [91]:
values = [int(i) for i in df['Company Age'] if i != '?']
mean = int((sum(values))/len(values))

for j, i in enumerate(df['Company Age']):
    if i == '?':
        df['Company Age'][j] = mean
        
df.drop('Founded', axis=1, inplace=True)


C:\ProgramData\Anaconda\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Transform into a downcast integer type to save memory once again

In [92]:
df['Company Age'] = pd.to_numeric(df['Company Age'], downcast='integer')

-------------------------------------------------------------------------------------------------------------------------------

# Chapter 2 
<a id='Chapter 2'></a>

Get the words we want to filter for in job title by first getting all the individual words in the 'job title' column and then filter and keep only the words with 15 occurences, then do some manaual cleaning.
The variable 'counts' contains all the individual words with the corresponding frequency associated with each word in a python  dictionary fashion whilst the 'wordsGreaterThan15' variable contains the words with greater than 15 occurences (minus/plus a few that were added and removed manually) in a python list fashion.

In [93]:
unwantedchar = '[,]./!{}()>?<|\&^*£$%^-_+='
counts = {}
text = ''
wordsGreaterThan15 = []



for j, i in enumerate(df['Job Title']):
    i = i.translate({ord(h): None for h in unwantedchar}) #remove unwanted chars in each value of job title
    i = i.lower() # make each value of job title lowercase
    df['Job Title'][j] = i # assign cleaned up value of Job Title back to relevant row 
    for j in i:
        if j != '' and j != ' ':
            text += j
        else:
            if text not in counts.keys():
                counts[text] = 1
                text = ''
            else:
                counts[text] += 1
                text = ''
sorted_counts = sorted(counts.items(), key=lambda x: x[1], reverse=True) #sort 'counts' with the key with largest value showing first and so forth

# iterate over the keys in 'counts' and check their corresponding value; if greater than 15 append to 'wordsGreaterThan15'

for i in counts.keys():
    if counts[i] >= 15:
        wordsGreaterThan15.append(i)

# delete words which are not useful so we do not to filter for them
        
del wordsGreaterThan15[2]
del wordsGreaterThan15[3]
del wordsGreaterThan15[-1]
del wordsGreaterThan15[8]
del wordsGreaterThan15[10]

# add in a few relevant words

wordsGreaterThan15.append('sr')
wordsGreaterThan15.append('jr')
wordsGreaterThan15.append('junior')
wordsGreaterThan15.append('entry')
wordsGreaterThan15.append('eng')
wordsGreaterThan15.append('engineering')
wordsGreaterThan15.append('remote')
wordsGreaterThan15.append('graduate')
wordsGreaterThan15.append('intern')
wordsGreaterThan15.append('fluency')
wordsGreaterThan15.append('fluent')

# add in skill words to filter for, this will be useful later 

skillwords = []
_ = [] # temporary, needed to avoid recursion, we will delete it once done to save memory

wordsGreaterThan15.append('python')
skillwords.append('python')
#wordsGreaterThan15.append('r')
skillwords.append('r')
wordsGreaterThan15.append('sas')
skillwords.append('sas')
wordsGreaterThan15.append('sap')
skillwords.append('sap')
wordsGreaterThan15.append('excel')
skillwords.append('excel')
wordsGreaterThan15.append('sql')
skillwords.append('sql')
wordsGreaterThan15.append('tableau')
skillwords.append('tableau')
wordsGreaterThan15.append('power')
skillwords.append('power')
wordsGreaterThan15.append('aws')
skillwords.append('aws')
wordsGreaterThan15.append('azure')
skillwords.append('azure')
wordsGreaterThan15.append('gcp')
skillwords.append('gcp')
wordsGreaterThan15.append('cloud')
skillwords.append('cloud')
wordsGreaterThan15.append('big')
skillwords.append('big')
    
wordsGreaterThan15 = sorted(wordsGreaterThan15)
print(f'wordsGreaterThan15 = {wordsGreaterThan15}')

for i in skillwords:
    _.append(i[0].upper() + i[1:])
    _.append(i.upper())
    
for i in _:
    skillwords.append(i)
    
del _
skillwords = sorted(skillwords)
print(f'skillwords = {skillwords}')

C:\ProgramData\Anaconda\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()


wordsGreaterThan15 = ['analyst', 'analystbusiness', 'analystdata', 'analystfinancial', 'analysthealthcare', 'analystjunior', 'analystlead', 'analystsenior', 'analystsr', 'analytics', 'analyticsdata', 'aws', 'azure', 'bi', 'big', 'business', 'cloud', 'data', 'eng', 'engineering', 'entry', 'excel', 'financial', 'fluency', 'fluent', 'gcp', 'governance', 'graduate', 'healthcare', 'intelligence', 'intern', 'jr', 'junior', 'lead', 'management', 'manager', 'master', 'operations', 'power', 'python', 'quality', 'remote', 'reporting', 'sap', 'sas', 'science', 'security', 'senior', 'sql', 'sr', 'systems', 'tableau', 'warehouse']
skillwords = ['AWS', 'AZURE', 'Aws', 'Azure', 'BIG', 'Big', 'CLOUD', 'Cloud', 'EXCEL', 'Excel', 'GCP', 'Gcp', 'POWER', 'PYTHON', 'Power', 'Python', 'R', 'R', 'SAP', 'SAS', 'SQL', 'Sap', 'Sas', 'Sql', 'TABLEAU', 'Tableau', 'aws', 'azure', 'big', 'cloud', 'excel', 'gcp', 'power', 'python', 'r', 'sap', 'sas', 'sql', 'tableau']


Filter through the individual job titles using the words in the list of viable words we have made (variable wordsGreaterThan15). Thus the new job titles should only contain one of those words else they will be blank.

In [95]:
wordsOfJobTitle = []
rcount = 0
blankcount = 0

for j, i in enumerate(df['Job Title']):
    wordsOfJobTitle.append([])
    #print(df['Job Title'][j])
    for letter in i:
        if letter != ' ':
            individualword += letter
        else:
            for _ in wordsGreaterThan15:
                if individualword == _:
                    wordsOfJobTitle[j].append(individualword)
                    individualword = ''
                elif individualword == 'r':
                    wordsOfJobTitle[j].append(individualword)
                    rcount += 1
                    individualword = ''
            individualword = ''
    df['Job Title'][j] = ''
    for index in range(len(wordsOfJobTitle[j])):
        df['Job Title'][j] += f'{wordsOfJobTitle[j][index]} '
    if df['Job Title'][j] == '' or df['Job Title'][j] == ' ':
        blankcount += 1
        
print(f"we've had {rcount} mentions of R!")
print(f"There are {blankcount} blanks now")
      
    

C:\ProgramData\Anaconda\lib\site-packages\ipykernel_launcher.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
C:\ProgramData\Anaconda\lib\site-packages\ipykernel_launcher.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


we've had 0 mentions of R!
there are 190 blanks now


Remove all the rows with a blank job title

In [99]:
for j, i in enumerate(df['Job Title']):
    if i == '' or i == ' ':
        df.drop([j], inplace=True)

Since we are going through words in a column quite often it's worthwhile to create a function to call instead of writing the process over and over again (probably should have done this earlier)

In [103]:
def go_through_words_in_column(column):
    
    listOfWords = []
    temp = ''
    for j, cell in enumerate(column):
        listOfWords.append([])
        for letter in cell:
            if letter != '' or letter != ' ':
                temp += letter
            else:
                listOfWords[j].append(temp)
                temp = ''
                
                
    return listOfWords